jans-cedarling/cedarling/src/jwt/service.rs

```diff
@@ -0,0 +1,130 @@
+/// and without validation, for cases where the token's validity is not required.
+#[derive(Clone)]
+pub struct JwtService {
+    config: Arc<JwtConfig>,
+    key_service: Option<Arc<dyn KeyService>>,
+}
+
+impl JwtService {
+    /// Creates a new instance of `JwtService`.
+    ///
+    /// If `CEDARLING_JWT_VALIDATION` is set to `Enabled` in the
+    /// bootstrap properties `key_service` cannot not be `None`.
+    pub fn new(config: JwtConfig, key_service: Option<Arc<dyn KeyService>>) -> Self {
+        Self {
+            config: Arc::new(config),
+            key_service,
+        }
+    }
+
+    /// Decodes a JWT, optionally validating it based on the configuration settings.
+    pub fn decode<T: serde::de::DeserializeOwned>(&self, jwt: &str) -> Result<T, Error> {
+        match self.config.as_ref() {
+            JwtConfig::Disabled => decode_jwt_without_validation(jwt),
+            JwtConfig::Enabled {
+                signature_algorithms,
+            } => self.decode_and_validate_jwt(jwt, signature_algorithms),
+        }
+    }
+
+    /// Decodes a JWT and returns the `TokenPayload`.
+    pub fn decode_token_data(&self, jwt: &str) -> Result<TokenPayload, Error> {
+        self.decode(jwt)
+    }
+
+    /// Decodes and validates a JWT using the specified algorithms and keys.
+    fn decode_and_validate_jwt<T: serde::de::DeserializeOwned>(
+        &self,
+        jwt: &str,
+        signature_algorithms: &Vec<String>,
+    ) -> Result<T, Error> {
+        let header = extract_jwt_header(jwt).map_err(|e| Error::ParsingError(e))?;
+
+        // Automatically reject unsupported algorithms
+        let alg_string = alg_to_string(header.alg);
+        if !signature_algorithms.contains(&alg_string) {
+            return Err(Error::UnsupportedAlgorithm);
+        }
+
+        let validator = Validation::new(header.alg);
+
+        // fetch decoding key from the KeyService
+        let key_service = self
+            .key_service
+            .as_ref()
+            .ok_or_else(|| Error::KeyServiceNotFound)?;
+        let key = key_service
+            .get_key(
+                &header
+                    .kid
+                    .ok_or(Error::MissingRequiredHeader("kid".to_string()))?,
+            )
+            .map_err(|e| Error::MissingKey(e))?;
+
+        let claims = decode::<T>(&jwt, &key, &validator)
+            .map_err(|e| Error::ValidationError(e))?
+            .claims;
+
+        Ok(claims)
+    }
+}
+
+/// Extracts the JWT header from the given JWT string without validation.
+fn extract_jwt_header(jwt: &str) -> Result<Header, jsonwebtoken::errors::Error> {
+    let header = decode_header(jwt)?;
+    Ok(header)
+}
+
+/// Decodes a JWT without performing any validation.
+fn decode_jwt_without_validation<T: serde::de::DeserializeOwned>(jwt: &str) -> Result<T, Error> {
+    let header = extract_jwt_header(jwt).map_err(|e| Error::ParsingError(e))?;
+
```

Comment:
we can just pass the error

```suggestion
    let header = extract_jwt_header(jwt).map_err(Error::ParsingError)?;
```

Reply:
This function no longer exists. just call `JwtService::decode(jwt)` from now on.

`JwtService` should now be smart enough to know what to do based on the passed in `JwtConfig`.